Simulate trigger channels for RNO-G (simplified)

fschlueter opened the pull request:
As the name suggests this PR is an evolution or devolution of #743. In particular it reverts the changes related to not adding the cable delay in the `eFieldToVoltageConverter` module. Reasons for this are discussed in the PR #743 but I am failing a technical implementation of a proper solution ....

For RNO-G this should not be a serious problem (for the moment) hopefully, since, I) We do not simulate coherent noise, and II) the trigger channels do not have a different relative cable delay to each other than the daq channels (at least no one which we are currently describing in the database).

fschlueter commented:
hallelujah!

fschlueter commented on NuRadioReco/modules/ARA/hardwareResponseIncorporator.py:
```diff
@@ -46,9 +47,18 @@ def run(self, evt, station, det, sim_to_data=False):
+                )
                 channel.set_frequency_spectrum(trace_before_system_fft, channel.get_sampling_rate())
 
+        if not sim_to_data:
+            # Subtraces the cable delay. For `sim_to_data=True`, the cable delay is added
+            # in the efieldToVoltageConverter or with the channelCableDelayAdder
+            # (if efieldToVoltageConverterPerEfield was used).
+            signal_processing.add_cable_delay(station, det, sim_to_data=False, logger=self.logger)
+
```
This is something I added to make this module (the ARA hardwareResponseIncorporator) more similar to the once for ARIANNA and RNO-G. I could remove it again to not change the functionality of the module?

anelles replied:
I don't have strong feelings. I don't think a lot of people are using this actively. I would be ok, to change it.

anelles commented on NuRadioMC/simulation/simulation.py:
```diff
@@ -562,7 +562,7 @@ def apply_det_response(evt, det, config,
         # start detector simulation
-        efieldToVoltageConverter.run(evt, station, det, channel_ids=channel_ids)  # convolve efield with antenna pattern
+        efieldToVoltageConverter.run(evt, station, det, channel_ids=channel_ids)  # convolve efield with antenna pattern and adds cable delay!
         # downsample trace to internal simulation sampling rate (the efieldToVoltageConverter upsamples the trace to
```
If you look at the two doc strings, it seems confusing. eFieldToVoltageConverter: Does it or does it not add the cable delay? We should be very explicit here now.

fschlueter replied:
The `efieldToVoltageConverter` does add the cable delay, the `efieldToVoltageConverterPEREFIELD` does not. I know ... confusing.

anelles commented on NuRadioReco/detector/response.py:
```diff
@@ -98,6 +99,8 @@ def __init__(self, frequency, y, y_unit, time_delay=0, weight=1,
             raise ValueError("Data for response incomplete, detected \"None\"")
 
+        # print(name, y_unit[1], y[1])
+
```
Can we loose this please?

fschlueter replied:
Woooow. Did expect anyone to catch that. I did not want to run the ci for this but of course I can.

anelles replied:
I can read ... 😇

sjoerd-bouma replied:
github hack - if you put [ci skip] in your commit message it doesn't run the ci!

anelles commented on NuRadioReco/detector/response.py:
```diff
@@ -418,13 +497,14 @@ def _calculate_time_delay(self):
-        # fit the unwrapped phase with a linear function
-        popt = np.polyfit(freqs, np.unwrap(phase), 1)
-        time_delay2 = -popt[0] / (2 * np.pi)
+        # This alternative calculation is only meaningful if group delay is ~ constant over the whole frequency range (which is the case for most cables)
```
So you decided to throw it out? I don't follow the logic. It is the case, so we don't calculate it to check for deviations?

fschlueter replied:
This alternative calculation (which is only well motivated for cable) was only used as a cross-check and caused a lot of warnings. Furthermore, the calculation itself is not used for physics results but only for consistency checks.

anelles replied:
Ok. I just wanted to understand why someone once thought that this was useful and now we simply remove it.

fschlueter replied:
The one was me :D

sjoerd-bouma commented on NuRadioReco/utilities/signal_processing.py:
```diff
@@ -27,3 +31,50 @@ def half_hann_window(length, half_percent=None, hann_window_length=None):
+
+def add_cable_delay(station, det, sim_to_data=None, trigger=False, logger=None):
+    """
+    Add or subtract cable delay by modifying the `trace_start_time`.
+
+    Parameters
+    ----------
+    station: Station
+        The station to add the cable delay to.
+
+    det: Detector
+        The detector description
+
+    trigger: bool
+        If True, take the time delay from the trigger channel response.
+        Only possible if `det` is of type `rnog_detector.Detector`. (Default: False)
+
+    logger: logging.Logger, default=None
+        If set, use `logger.debug(..)` to log the cable delay.
```
docstring formatting
- use double backticks for single-spaced in rst. Single backticks are interpreted as references
- Can you add a `See Also` / link to the CableDelayAdder class?

fschlueter replied:
Like this?

sjoerd-bouma commented on NuRadioMC/simulation/simulation.py:
```diff
@@ -1622,6 +1632,12 @@ def run(self):
                         channel[chp.Vrms_NuRadioMC_simulation] = self._Vrms_per_channel[station_id][channel.get_id()]
                         channel[chp.bandwidth_NuRadioMC_simulation] = self._integrated_channel_response[station_id][channel.get_id()]
 
+                        # Always store this, even when no "extra" trigger channels are simulated
+                        if self.__trigger_channel_ids is not None and channel.get_id() in self.__trigger_channel_ids and channel.get_id() in self._Vrms_per_trigger_channel[station_id]:
+                            channel[chp.Vrms_trigger_NuRadioMC_simulation] = self._Vrms_per_trigger_channel[station_id][channel.get_id()]
+                        else:
+                            channel[chp.Vrms_trigger_NuRadioMC_simulation] = 0
```
I'm slightly confused here... if there are no 'additional trigger channels' RNO-G style, this parameter is still set, but it's set to 0? I would have thought it makes more sense to either default it to the Vrms of the 'normal' channel if there is no additional trigger channel associated, or to just not set it. But probably I'm missing something?

fschlueter replied:
I did not want to set it to the same value as the normal channel and I think it made it easier to just set it to zero in case there are no explicit trigger channels.

fschlueter replied:
Fine?

sjoerd-bouma replied:
Unless I'm still missing something I would suggest not setting this. Having it set to an 'invalid' value like 0 will either lead to problems anyway (e.g. division by 0 for SNR calculations) or lead to wrong conclusions (e.g. the user thinks the trigger channels were simulated without noise if there are no additional trigger channels).

fschlueter replied:
I am pretty sure I did it for a (good) reason. But lets see. I can take it of an run it again.

fschlueter replied:
Okay, I am now only storing the attribute for the trigger channels.

anelles commented:
> Also note, currently only two trigger modules are modified to use the trigger channels (if present) the highlow and simple threshold triggers. All other triggers are not yet converted.

Should we add a warning in those?

sjoerd-bouma commented on NuRadioReco/framework/parameters.py:
```diff
@@ -62,7 +62,7 @@ class channelParameters(Enum):
     Vrms_NuRadioMC_simulation = 19  #: the noise rms used in the MC simulation
     bandwidth_NuRadioMC_simulation = 20  #: the integrated channel response (=bandwidth for signal chains without amplification) used in the MC simulation
-
+    Vrms_trigger_NuRadioMC_simulation = 21  #: the noise rms of the trigger channels (otional) used in the MC simulation
```
typo (optional).

Also, I don't know if this is the best place for it, but for non-RNO-G experts, it might be nice to know why this is optional, and when it is/isn't used?

sjoerd-bouma commented on NuRadioReco/utilities/signal_processing.py:
```diff
@@ -27,3 +31,54 @@ def half_hann_window(length, half_percent=None, hann_window_length=None):
     return half_hann_widow
+
+
+def add_cable_delay(station, det, sim_to_data=None, trigger=False, logger=None):
+    """
+    Add or subtract cable delay by modifying the `trace_start_time`.
+
+    Parameters
+    ----------
+    station: Station
+        The station to add the cable delay to.
+
+    det: Detector
+        The detector description
+
+    trigger: bool
+        If True, take the time delay from the trigger channel response.
+        Only possible if ``det`` is of type `rnog_detector.Detector`. (Default: False)
+
+    logger: logging.Logger, default=None
+        If set, use ``logger.debug(..)`` to log the cable delay.
+
+    See Also
+    --------
+    `NuRadioReco.modules.channelAddCableDelay.channelAddCableDelay`
+    """
```
I just realized this is just `channelAddCableDelay.run`, with the added option to apply delays to the additional trigger channels only... I guess the code isn't duplicated because you made channelAddCableDelay wrap this function, but I'm not sure I understand why you didn't just add the `trigger` flag to `channelAddCableDelay` so it's all in one place.

I would probably be a happier man if I could understand your motivation for this! But I can live with this, at least the code is in one place and they link to each other.

fschlueter replied:
Because I am using this helper function also in the 3 hardwareResponseIncorperator modules. We could call the cable delay adder module within this modules but this would produce quite some overhead (boiler) code which is not needed with this solution.

sjoerd-bouma replied:
Am I happy? Not that happy. But it's a good enough reason. (It's true a function makes more sense than a class in this case as channelAddCableDelay doesn't use `self` at all... one could make the two completely identical (apart from the unused `event` in the signature) by just adding a `@staticmethod` decorator and using that. But that would not necessarily be a much cleaner solution and I think I really will be able to live with this.)

sjoerd-bouma commented:
I think just two more small comments and then I'm happy enough:
- As Anna mentioned, it's not ideal that only some of the triggers use `iter_trigger_channels` and the others use `iter_channels`. Do you think you could change this so all triggers use the trigger_channels? Otherwise I fear this becomes either one of those 'to-do' items that might be forgotten and result in unintuitive trigger behaviour in the future.
- Please add this new feature/change to the changelog :)
- (3/2) fix the merge conflict :))

sjoerd-bouma commented:
(4/2) I ran the example but wasn't expecting any more issues... but I'm getting an error when trying to read the resulting .nur file:
```
KeyError                                  Traceback (most recent call last)
Cell In[2], [line 1](vscode-notebook-cell:?execution_count=2&line=1)
----> [1](vscode-notebook-cell:?execution_count=2&line=1) for evt in reader.run():
      [2](vscode-notebook-cell:?execution_count=2&line=2)     pass

File ~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/NuRadioRecoio.py:326, in NuRadioRecoio.get_events(self)
    [324](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/NuRadioRecoio.py:324) self._current_file_id = 0
    [325](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/NuRadioRecoio.py:325) self._get_file(self._current_file_id).seek(12)  # skip file header
--> [326](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/NuRadioRecoio.py:326) for event in self.__iter_events(self):
    [327](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/NuRadioRecoio.py:327)     self._current_event_id = event.get_id()
    [328](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/NuRadioRecoio.py:328)     self._current_run_number = event.get_run_number()

File ~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/event_parser_factory.py:202, in iter_events_function.<locals>.iter_events_2_2(self)
    [200](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/event_parser_factory.py:200)     evtstr = self._get_file(self._current_file_id).read(bytes_to_read)
    [201](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/event_parser_factory.py:201)     event = NuRadioReco.framework.event.Event(0, 0)
--> [202](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/event_parser_factory.py:202)     event.deserialize(evtstr)
    [203](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/event_parser_factory.py:203)     yield event
    [204](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/modules/io/event_parser_factory.py:204) elif object_type == 1 or object_type == 2:

File ~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/event.py:557, in Event.deserialize(self, data_pkl)
    [555](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/event.py:555) for station_pkl in data['stations']:
    [556](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/event.py:556)     station = NuRadioReco.framework.station.Station(0)
--> [557](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/event.py:557)     station.deserialize(station_pkl)
    [558](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/event.py:558)     self.set_station(station)
    [559](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/event.py:559) if 'showers' in data.keys():

File ~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/station.py:238, in Station.deserialize(self, data_pkl)
    [236](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/station.py:236) else:
    [237](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/station.py:237)     self.__sim_station = NuRadioReco.framework.sim_station.SimStation(None)
--> [238](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/station.py:238)     self.__sim_station.deserialize(data['sim_station'])
    [239](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/station.py:239) for channel_pkl in data['channels']:
    [240](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/station.py:240)     channel = NuRadioReco.framework.channel.Channel(0)

File ~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_station.py:161, in SimStation.deserialize(self, data_pkl)
    [159](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_station.py:159) for channel_pkl in data['channels']:
    [160](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_station.py:160)     channel = NuRadioReco.framework.sim_channel.SimChannel(0, 0, 0)
--> [161](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_station.py:161)     channel.deserialize(channel_pkl)
    [162](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_station.py:162)     self.add_channel(channel)

File ~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_channel.py:64, in SimChannel.deserialize(self, data_pkl)
     [62](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_channel.py:62) def deserialize(self, data_pkl):
     [63](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_channel.py:63)     data = pickle.loads(data_pkl)
---> [64](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_channel.py:64)     NuRadioReco.framework.channel.Channel.deserialize(self, data['channel'])
     [65](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_channel.py:65)     self._shower_id = data['shower_id']
     [66](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/sim_channel.py:66)     self._ray_tracing_id = data['ray_tracing_id']

File ~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:137, in Channel.deserialize(self, data_pkl)
    [135](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:135) if trigger_channel_pkl is not None:
    [136](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:136)     trigger_channel = Channel(None)
--> [137](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:137)     trigger_channel.deserialize(trigger_channel_pkl)
    [138](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:138)     self._trigger_channel = trigger_channel
    [139](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:139) else:

File ~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:126, in Channel.deserialize(self, data_pkl)
    [124](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:124) def deserialize(self, data_pkl):
    [125](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:125)     data = pickle.loads(data_pkl)
--> [126](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:126)     if data['base_trace'] is not None:
    [127](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:127)         NuRadioReco.framework.base_trace.BaseTrace.deserialize(self, data['base_trace'])
    [129](https://file+.vscode-resource.vscode-cdn.net/home/sb/Python/scratch/debugging/~/Python/software/nu-radio/NuRadioMC/NuRadioReco/framework/channel.py:129)     self._parameters = NuRadioReco.framework.parameter_serialization.deserialize(data['parameters'], parameters.channelParameters)

KeyError: 'base_trace'
```

sjoerd-bouma commented:
> On it! And my goal, of course, is to make you very happy Sjoerd!

I admire your ambition! But I fear the problem is me, not you ;)

fschlueter commented:
> > On it! And my goal, of course, is to make you very happy Sjoerd!
> 
> I admire your ambition! But I fear the problem is me, not you ;)

I like a good challenge :D

fschlueter commented:
@sjoerd-bouma took me a while to understand the problem and than find the best solution. The issue was that also `sim_channel` could get a trigger_channel. Which was not envisioned by me (the `isinstance` check does not work for inherited classes). I finally decided to change the code such that sim_channel are never getting trigger channels (rather than allowing them to deserialize those which was the cause of the issue)

fschlueter commented:
I think I addressed all your points @sjoerd-bouma

fschlueter commented:
@sjoerd-bouma, do you think we can continue the review?

sjoerd-bouma commented on NuRadioMC/simulation/simulation.py:
```diff
@@ -1666,6 +1688,10 @@ def run(self):
                         channel[chp.Vrms_NuRadioMC_simulation] = self._Vrms_per_channel[station_id][channel.get_id()]
                         channel[chp.bandwidth_NuRadioMC_simulation] = self._integrated_channel_response[station_id][channel.get_id()]
 
+                        # Always store this, even when no "extra" trigger channels are simulated
```
maybe remove this comment also now that it no longer applies

sjoerd-bouma commented on NuRadioReco/detector/RNO_G/rnog_detector.py:
```diff
@@ -1255,6 +1278,45 @@ def get_site_coordinates(self, station_id=None):
+        return self.__db
+
+
+    def get_component(self, collection="coax_cable", component="daq_drab_flower_2024_avg"):
```
If this is the 'wrapper' around `get_component_data` you mentioned in #773 , with the intention that it should be used - please add a docstring; I'm happy for you to copy/update the documentation I added to explain this in #773 (maybe we merge that first?)

sjoerd-bouma commented on NuRadioReco/framework/station.py:
```diff
@@ -94,9 +94,36 @@ def iter_channels(self, use_channels=None, sorted=False):
         for channel_id in channel_ids:
             yield self.get_channel(channel_id)
 
+    def iter_trigger_channels(self, use_channels=None):
```
docstring? (I know iter_channels doesn't have one either but maybe it helps to distinguish the two?)

sjoerd-bouma commented on NuRadioReco/modules/RNO_G/hardwareResponseIncorporator.py:
```diff
@@ -1,17 +1,20 @@
     """
-    Incorporates the gain and phase induced by the RNO-G hardware.
+    Incorporates the compex response of the RNO-G hardware. The response is obtained from the detector description.
+    The response is applied in the frequency domain. This module also adds the cable delay to the channels.
```
This is no longer true, right? (It doesn't add the cable delays?)

sjoerd-bouma commented on NuRadioReco/modules/RNO_G/hardwareResponseIncorporator.py:
```diff
@@ -76,15 +84,18 @@ def get_filter(self, frequencies, station_id, channel_id, det,
             Note: The adjustment to the minimal gain is NOT visible when getting the amp response from
             ``analog_components.get_amplifier_response()``
 
+        is_trigger_channel: bool
+            Use trigger channel response instead. Only relevant for RNO-G. (Default: False)
```
Doesn't match the parameter name. Is it worth just using `trigger` to agree with the name in the detector module?

sjoerd-bouma reviewed (commented):
I

sjoerd-bouma commented:
> @sjoerd-bouma took me a while to understand the problem and than find the best solution. The issue was that also `sim_channel` could get a trigger_channel. Which was not envisioned by me (the `isinstance` check does not work for inherited classes). I finally decided to change the code such that sim_channel are never getting trigger channels (rather than allowing them to deserialize those which was the cause of the issue)

I need to look a bit deeper because a) I didn't look deeply enough at first to understand what was causing this issue and b) it's not obvious to me what you've changed in the last 10 commits that fixed it (though I can confirm I can now read the .nur file produced by the example). I'm not sure why you don't want sim_channels to have trigger_channels? I thought they were generally intended to have a noiseless, MC truth record of the trace, so I'm not so sure I understand why you don't want to store the trigger_channel in the same way.

sjoerd-bouma commented:
Also, it looks like the trigger readout windows are a bit funky in the example, at least when I ran it all the pulses in the phased array were very close to the start.

fschlueter commented:
@sjoerd-bouma I think we can discuss this in a separate PR (whether sim_channels are suppose to have trigger_channels as well). I can easily revert these changes again. But I would appreciate if we could do that in an new PR. What I did to fix the issue: I changed the `isinstance(...)` to `type(...) ==` and added in the hardwareResponseIncorperator an extra condition that the station object has to be a `Station` (and not be a `SimStation`).

fschlueter commented:
> Also, it looks like the trigger readout windows are a bit funky in the example, at least when I ran it all the pulses in the phased array were very close to the start.

I would also argue that this could be tackled in a new PR, anyway we want to get rid of the trigger_readout window module in its current form? Although if you insist I can have a look,

fschlueter commented:
@sjoerd-bouma I hopefully resolved all your comments.

sjoerd-bouma commented:
> > Also, it looks like the trigger readout windows are a bit funky in the example, at least when I ran it all the pulses in the phased array were very close to the start.
> 
> I would also argue that this could be tackled in a new PR, anyway we want to get rid of the trigger_readout window module in its current form? Although if you insist I can have a look,

I checked (our trigger code is indeed not very transparent...) and several modules deep found that the issue is that the default value for `pre_trigger_time` in the `HighLowTrigger` is 55 ns. So that would be why. This should probably be looked at (`trigger.highLowThreshold.triggerSimulator` doesn't even allow to set the pre_trigger_time, so you're always stuck with this default), but I agree that's not an issue for this PR.
